datafusion/core/src/datasource/function.rs

```diff
@@ -0,0 +1,60 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! A table that uses a function to generate data
+
+use std::sync::Arc;
+
+use super::{streaming::StreamingTable, TableProvider};
+use arrow::datatypes::SchemaRef;
+use datafusion_common::Result;
+use datafusion_physical_plan::streaming::PartitionStream;
+
+/// Table Function implementation
+pub type TableFunctionImplementation = Arc<
+    dyn Fn(&[String]) -> Result<(SchemaRef, Vec<Arc<dyn PartitionStream>>)> + Send + Sync,
+>;
+
+/// A table that uses a function to generate data
+pub struct TableFunction {
+    /// Name of the table function
+    name: String,
+    /// Function implementation
+    fun: TableFunctionImplementation,
+}
+
+impl TableFunction {
+    /// Create a new table function
+    pub fn new(name: String, fun: TableFunctionImplementation) -> Self {
+        Self { name, fun }
+    }
+
+    /// Get the name of the table function
+    pub fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    /// Get the function implementation and generate a table
+    pub fn create_table_provider(
+        &self,
+        args: &[String],
```

Comment:
Instead of taking strings what would you think about taking `Expr` instead (which would allow for more sophisticated things like passing in `now()` as sugested by @yukkit  in https://github.com/apache/arrow-datafusion/issues/7926#issuecomment-1812390523)?

Reply:
Love this idea too! I'll try to figure out iff we can find a way to evaluate `Expr` correctly!

Reply:
@alamb  PTAL I have made`Exprs` and provider inputs and it looks just fine!

but we still have some problems to solve, you can check TODO in the latest commit.

- [ ] how to evaluate expr simpler? 
    - for now, I left it in `fn scan` to construct `Expr` -> `LogicalPlan` -> `PhysicalPlan` and evaluate it then, but I think we can make it simpler or elsewhere to evaluate `Exprs`.

Reply:
>  how to evaluate expr simpler?

you can use [`ExprSimplifier::simplify`](https://docs.rs/datafusion/latest/datafusion/optimizer/simplify_expressions/struct.ExprSimplifier.html#method.simplify) to do this. 

Here is an example https://github.com/apache/arrow-datafusion/blob/06bbe1298fa8aa042b6a6462e55b2890969d884a/datafusion-examples/examples/expr_api.rs#L56

I added a commit to  https://github.com/Veeupup/arrow-datafusion/pull/1 showing how to make it work: https://github.com/Veeupup/arrow-datafusion/pull/1/commits/2df01dbae67fe60746901d07f1729e1d2d8e6bd6

Reply:
> Love this idea too! I'll try to figure out if we can find a way to evaluate `Expr` correctly!

BTW since I agree evaluating expressions is non trivial, I added some additional documentation and examples of how to do so here: https://github.com/apache/arrow-datafusion/pull/8377